server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,8 +245,11 @@ private void performOnReplicas(
 
         final ShardRouting primaryRouting = primary.routingEntry();
 
-        for (final ShardRouting shard : replicationGroup.getReplicationTargets()) {
-            if (shard.isSameAllocation(primaryRouting) == false) {
+        for (final ReplicationAwareShardRouting shardRouting : replicationGroup.getReplicationTargets()) {
+            ShardRouting shard = shardRouting.getShardRouting();
+            // TODO - Add condition of underlying action being replicated regardless i.e. shard bulk and publish checkpoint action
+            if (!shard.isSameAllocation(primaryRouting)
+                && (shardRouting.isReplicated() || (shardRouting.isRemoteTranslogEnabled() && forceReplicationIfRemoteTranslogEnabled))) {
```

Comment:
Why can't we implement a `ReplicasProxy` which doesn't send request to the replicas?

Reply:
There are multiple cases we need to support - 
- primary-primary peer recovery where the replication should work as is. 
- Ability to support certain transport replication actions like publish checkpoint action.
- and then No-Op replication has to be supported using existing transport shard bulk action.

Let me see if all of these cases can be supported with replicas proxy with keeping the code readable, maintainable and extensible.

Reply:
@Bukhtawar  - When we do not want the replication calls to fan out, the replica proxy comes in only at this point.
https://github.com/opensearch-project/OpenSearch/blob/0210b76dc70538370d2d272a72b7ab99dc18f0bc/server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java#L312

and the below piece of code would get executed as well, while this is not required in case of request to not fan out to replication group - 
https://github.com/opensearch-project/OpenSearch/blob/0210b76dc70538370d2d272a72b7ab99dc18f0bc/server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java#L243-L258

Reply:
Sorry couldn't follow you, but are you suggesting it might not be possible to create a proxy to control replication

Reply:
It is possible, but I think it does 2 things which can be avoided -
- Executes some code which is totally unnecessary (mentioned in my earlier comment)
- It would require create redundant replicaProxy on each of the TransportReplicationAction. IMO, this would just add lot of conditional logic in each of the TransportReplicationAction which the solution is this PR is handling in just one place itself.

Reply:
Sorry but what prevents reuse since you mentioned we need redundant code 
Also why can't we wire replication actions to use a flavor of replica proxy

Reply:
We would need to then handle the execution of unnecessary code that would get executed if we trigger the performOnReplicas method which then hits the replicaProxy.peform method. Not just that, but the added logic for each of transport replication actions. 
Do you see any reason that why we must let the call pass through even with the controls that has been put up (allowing certain transport replication to proceed - publish checkpoint action & transport shard bulk action)?